packages/stack/compiler/test/compiler.spec.js

```diff
@@ -0,0 +1,60 @@
+import sinon from 'sinon';
+import assert from 'assert';
+import { File, Types } from 'embark-utils';
+import { fakeEmbark } from 'embark-testing';
+import Compiler from '../src/';
+
+const { embark, plugins } = fakeEmbark();
+
+// Due to our `DAPP_PATH` dependency in `embark-utils` `dappPath()`, we need to
+// ensure that this environment variable is defined.
+process.env.DAPP_PATH = 'something';
+
+describe('stack/compiler', () => {
+
+  let compiler, doneCb;
+
+  beforeEach(() => {
+    compiler = new Compiler(embark, { plugins });
+  });
+
+  afterEach(() => {
+    embark.teardown();
+    sinon.restore();
+  });
+
+  test('it should use registered compiler', done => {
+    const fooCompiler = sinon.spy((files, options, cb) => cb(null, {
+      contractA: 'someResultA',
+      contractB: 'someResultB',
+      contractC: 'someResultC',
+    }));
+
+    embark.plugins.createPlugin('fooCompiler').registerCompiler('.foo', fooCompiler);
+
+    embark.events.request('compiler:contracts:compile', [
+      new File({filename: 'foo.foo', type: Types.dappFile, path: 'foo.foo'}),
+      new File({filename: 'foo2.foo', type: Types.dappFile, path: 'foo2.foo'}),
+      new File({filename: 'foo3.foo', type: Types.dappFile, path: 'foo3.foo'}),
+    ], () => {
+      assert(fooCompiler.called);
+      done();
+    });
+  });
+
+  test('should iterate over available compilers to find a match for a given source file', done => {
+    const fooCompiler = sinon.spy((files, options, cb) => cb(null, { fooContract: 'foo', }));
+    const barCompiler = sinon.spy((files, options, cb) => cb(null, { barContract: 'bar', }));
+
+    embark.plugins.createPlugin('fooCompiler').registerCompiler('.foo', fooCompiler);
+    embark.plugins.createPlugin('barCompiler').registerCompiler('.bar', barCompiler);
+
+    embark.events.request('compiler:contracts:compile', [
+      new File({filename: 'foo.bar', type: Types.dappFile, path: 'foo.bar'}),
+    ], () => {
+      assert(fooCompiler.notCalled);
+      assert(barCompiler.called);
+      done();
+    });
+  });
+});
```

Comment:
One test that is missing here is to check how Embark behaves when there's no matching compiler for a given source file. As per spec here: https://notes.status.im/zwVzwz_SRTCrhk6IsZzdvQ#Testing, we expect Embark to throw and error (and probably output a useful message to the user).

This doesn't seem to be like how Embark behaves at the moment. When there's no compiler registered that matches a given file extension, the compiler returns earlier without an error: https://github.com/embark-framework/embark/blob/183d9a05b546fc8ec512a9d510ddf8f63b43d611/packages/stack/compiler/src/index.ts#L84-L87

There is a case where we throw an error here: https://github.com/embark-framework/embark/blob/183d9a05b546fc8ec512a9d510ddf8f63b43d611/packages/stack/compiler/src/index.ts#L97 

This however only applies when there's version mismatch, not a compiler mismatch.

I'm not sure what the actual expected behaviour is now. To me it sounds like we have to change the compiler so it'll throw where it returns right now?

Reply:
So, we do have the warning. It's a the bottom: https://github.com/embark-framework/embark/blob/183d9a05b546fc8ec512a9d510ddf8f63b43d611/packages/stack/compiler/src/index.ts#L104

Thing is, we don't want to throw where you pointed out, because it just means that that specific compiler doesn't have any contract file it can compile.

What we can do, is throw at the end when we detect that some contracts have not been compiled or for the tests, just do the same check as the compiler module is doing to see if some of them are not compiled.

Reply:
> What we can do, is throw at the end when we detect that some contracts have not been compiled or for the tests, just do the same check as the compiler module is doing to see if some of them are not compiled.

That makes sense, thanks! I guess I'll go with doing the same check in the tests.